website/src/libs/api-v3/cron.js

```diff
@@ -0,0 +1,274 @@
+        // initially, and when we realised, we could not fix it because users are used to
+        // their Medium and Hard Dailies doing an Easy amount of damage from boss.
+        // Easy is task.priority = 1. Anything < 1 will be Trivial (0.1) or any future
+        // setting between Trivial and Easy.
+      }
+    }
+
+    task.history.push({
+      date: Number(new Date()),
+      value: task.value,
+    });
+    task.completed = false;
+
+    if (completed || scheduleMisses > 0) {
+      task.checklist.forEach(i => i.completed = true); // FIXME this should not happen for grey tasks unless they are completed
+    }
+  });
+
+  tasksByType.habits.forEach((task) => { // slowly reset 'onlies' value to 0
+    if (task.up === false || task.down === false) {
+      task.value = Math.abs(task.value) < 0.1 ? 0 : task.value = task.value / 2;
+    }
+  });
+
+  // Finished tallying
+  user.history.todos.push({date: now, value: todoTally});
+
+  // tally experience
+  let expTally = user.stats.exp;
+  let lvl = 0; // iterator
+  while (lvl < user.stats.lvl - 1) {
+    lvl++;
+    expTally += common.tnl(lvl);
+  }
+
+  user.history.exp.push({date: now, value: expTally});
+
+  // preen user history so that it doesn't become a performance problem
+  // also for subscribed users but differentyly
+  // premium subscribers can keep their full history.
+  preenUserHistory(user, tasksByType, user.preferences.timezoneOffset);
+
+  if (perfect) {
+    user.achievements.perfect++;
+    let lvlDiv2 = Math.ceil(common.capByLevel(user.stats.lvl) / 2);
+    user.stats.buffs = {
+      str: lvlDiv2,
+      int: lvlDiv2,
+      per: lvlDiv2,
+      con: lvlDiv2,
+      stealth: 0,
+      streaks: false,
+    };
+  } else {
+    user.stats.buffs = _.cloneDeep(clearBuffs);
+  }
+
+  // Add 10 MP, or 10% of max MP if that'd be more. Perform this after Perfect Day for maximum benefit
+  // Adjust for fraction of dailies completed
+  user.stats.mp += _.max([10, 0.1 * user._statsComputed.maxMP]) * dailyChecked / (dailyDueUnchecked + dailyChecked);
+  if (user.stats.mp > user._statsComputed.maxMP) user.stats.mp = user._statsComputed.maxMP;
+
+  if (dailyDueUnchecked === 0 && dailyChecked === 0) dailyChecked = 1;
+  user.stats.mp += _.max([10, 0.1 * user._statsComputed.maxMP]) * dailyChecked / (dailyDueUnchecked + dailyChecked);
+  if (user.stats.mp > user._statsComputed.maxMP) {
+    user.stats.mp = user._statsComputed.maxMP;
+  }
+
+  // After all is said and done, progress up user's effect on quest, return those values & reset the user's
+  let progress = user.party.quest.progress;
+  let _progress = _.cloneDeep(progress);
+  _.merge(progress, {down: 0, up: 0});
+  progress.collect = _.transform(progress.collect, (m, v, k) => m[k] = 0);
+
+  // Clean PMs - keep 200 for subscribers and 50 for free users
+  let maxPMs = 200;
+  let numberOfPMs = Object.keys(user.inbox.messages).length;
+  if (Object.keys(user.inbox.messages).length > maxPMs) {
```

Comment:
there's a duplicate call to Object.keys(user.inbox.messages).length. Also to be sure of not having problems with the special mongoose objects I'd do `user.inbox.message.toObject()` apply the changes to the new object and then replace user.inbox.messages with it